Grammar improvements

Mats-SX opened the pull request:
- Enforce valid clause structure in the grammar
- Add namespace part to procedure names

Also fixes a few syntax bugs in the TCK.

thobe commented on grammar/cypher.xml:
```diff
@@ -89,6 +84,51 @@
+    </alt>
+  </production>
+
+  <production name="StartingClause">
```
we probably want this inlined in railroad diagrams: `rr:inline="true"`.

thobe commented on grammar/cypher.xml:
```diff
@@ -89,6 +84,51 @@
+    </alt>
+  </production>
+
+  <production name="StartAndEndClause">
```
`rr:inline="true"`

thobe commented on grammar/cypher.xml:
```diff
@@ -89,6 +84,51 @@
+    </alt>
+  </production>
+
+  <production name="EndingClause">
```
`rr:inline="true"`

thobe commented on grammar/cypher.xml:
```diff
@@ -89,6 +84,51 @@
+    <alt>
+      <seq>
+        <non-terminal ref="StartingClause"/> &WS;
+        <repeat><non-terminal ref="Clause"/> &WS;</repeat>
```
I don't think we should allow _any_ clause in the middle of this sequence. For example once you hit `Return` it is over.

Mats-SX replied:
Right, this could be solved by excluding `RETURN` from `Clause`.

FylmTM commented:
Just a minor comment from my side:
- I really like how clauses are separated. It helps a lot, especially when you think about Cypher support in editors.

Even better if it can be improved as @thobe said.

Would love to see this changes merged.

thobe commented:
I'm thinking something like this (graphviz):


        init -> {LoadCSV Start Match Unwind Merge Create CreateUnique Foreach InQueryCall With Return}
        {Return InQueryCall Foreach Remove Delete Set CreateUnique Create Merge} -> done

        LoadCSV      -> {Start Match Unwind Merge Create CreateUnique Foreach InQueryCall With Return}
        Start        -> {Start Match Unwind Merge Create CreateUnique Set Delete Remove Foreach InQueryCall With Return}
        Match        -> {Match Unwind Merge Create CreateUnique Set Delete Remove Foreach InQueryCall With Return}
        Unwind       -> {Match Unwind Merge Create CreateUnique Set Delete Remove Foreach InQueryCall With Return}
        Merge        -> {Match Unwind Merge Create CreateUnique Set Delete Remove Foreach InQueryCall With Return}
        Create       -> {Create CreateUnique Set Delete Remove With Return}
        CreateUnique -> {Create CreateUnique Set Delete Remove With Return}
        Set          -> {Set Remove With Return}
        Delete       -> {Delete Remove With Return}
        Remove       -> {Remove With Return}
        Foreach      -> {Match Unwind Merge Create CreateUnique Set Delete Remove Foreach InQueryCall With Return}
        InQueryCall  -> {Match Unwind Merge Create CreateUnique Set Delete Remove Foreach InQueryCall With Return}
        With         -> {LoadCSV Start Match Unwind Merge Create CreateUnique Set Delete Remove Foreach InQueryCall With Return}

Although I don't know if these progressions are actually correct. But starting from such an exhaustive listing, I think it'll be easier to actually categorize the clauses and order them.

Mats-SX commented:
@thobe @FylmTM Updated with a more explicit and sophisticated clause structure. It splits queries into two categories: ones containing a `WITH` clause, and ones that do not. The observation is that the last query part always follows the same semantic rules, and a `WITH` clause kind of resets these conditions, with the exception of a few updating clauses that are no allowed to start a query (`SET`, `REMOVE`, `DELETE`).

I've added a new test where which takes invalid queries and validates that the Antlr grammar reports syntax errors. We can extend that with more test cases as we go.

I've removed the commits relating other changes (one pushed, one PR'd).

thobe commented on grammar/cypher.xml:
```diff
@@ -89,22 +85,107 @@
+    </alt>
+  </production>
+
+  <production name="ReadingClauseExceptWith" rr:inline="true">
```
I don't think `ExceptWith` adds anything useful to the name, arguably `WITH` doesn't do any reading anyhow ;)

Mats-SX replied:
I also cringed a bit when writing it -- but at that point there was already a group `ReadingClause`. Happy to remove that part now.

thobe commented on tools/grammar/src/test/java/org/opencypher/tools/grammar/Antlr4ParserTest.java:
```diff
@@ -67,12 +99,50 @@ public void investigateTokenStream() throws IOException
+            }
+//            else
+//            { break; }
+//        }
```
these commented out blocks make no sense
